EL5001 and EL5002 template/subs

JJL772 opened the pull request:
Support for EL5001/2 was added in #12, but that PR was missing the template/subs. This PR adds those, and fixes a bug with the number of registers being read out of the device.

Currently based on #36, so I'm leaving this as a draft until that is merged.

JJL772 committed:
ENH: support for EL5001 and EL5002 encoder terminals

JJL772 committed:
STY: apply clang-format

JJL772 committed:
CI test please ignore

klauer commented:
A CI job that checks that there are the correct set of template files/substitution files is probably overkill, right?

JJL772 commented:
Not a bad idea imo

JJL772 committed:
ENH: add CI check for terminals.json and db templates

JJL772 commented:
I've added a check for all db templates, and a schema check for terminals.json. Need to debug a problem with the mingw build before this is ready to merge though

JJL772 committed:
FIX: fix bit fields for mingw/windows builds

It seems like it inserts padding when there's space between the current
field and the next one, if they're differing types.

e.g. uint8_t a : 3; uint16_t b : 5; would produce a 2-byte struct even
though it looks like the bits should be packed. GCC and clang on Linux
handle this as expected.

JJL772 commented:
Hooray, the PDO safety macros caught compiler shenanigans. Ready to review now.

klauer commented on ek9000App/src/scripts/ci_check.py:
```diff
@@ -0,0 +1,115 @@
+
+import json
+import argparse
+import jsonschema
```
We probably should add a `requirements.txt` for these scripts somewhere.

I guess we just got lucky that GHA has jsonschema/lxml installed in its default Python environment? I wouldn't have expected it to have anything but the basics, so that surprised me a bit. In our typical CI pipelines, we use https://github.com/actions/setup-python for pip-based testing.

JJL772 replied:
Probably a good idea, totally slipped my mind!!

klauer commented on ek9000App/src/scripts/ci_check.py:
```diff
@@ -0,0 +1,115 @@
+
+def check_json(file: dict) -> bool:
+    print('Validating terminals.json...', end='')
+    jsonschema.validate(file, SCHEMA)
```
Just double-checking: `jsonschema.validate()` will raise on error: https://python-jsonschema.readthedocs.io/en/stable/validate/ so this is fine

JJL772 replied:
Yup

klauer reviewed (approved):
Changes LGTM - glad to see that the static checks caught issues ahead of time. 👍 

CI checker script is a nice addition, too!

JJL772 committed:
FIX: install python dependencies in CI
